external/fv3fit/fv3fit/emulation/transforms/factories.py

```diff
@@ -45,6 +49,83 @@ def build(self, sample: TensorDict) -> TensorTransform:
         return UnivariateTransform(self.source, self.to, self.transform)
 
 
+def reduce_std(x: tf.Tensor) -> tf.Tensor:
+    mean = tf.reduce_mean(x)
+    return tf.sqrt(tf.reduce_mean((x - mean) ** 2))
+
+
+def fit_conditional(
+    x: tf.Tensor, y: tf.Tensor, reduction: Callable[[tf.Tensor], tf.Tensor], bins: int,
+) -> Callable[[tf.Tensor], tf.Tensor]:
+    min = tf.reduce_min(x)
+    max = tf.reduce_max(x)
+    edges = tf.linspace(min, max, bins + 1)
+    values = groupby_bins(edges, x, y, reduction)
+
+    def interp(x: tf.Tensor) -> tf.Tensor:
+        return piecewise(edges[:-1], values, x)
+
+    return interp
+
+
+@dataclasses.dataclass
+class ConditionallyScaled(TransformFactory):
+    """Conditionally scaled transformation
+
+    Scales the output-input difference of ``field`` by conditional standard
+    deviation and mean::
+
+                  d field - E[d field|on]
+        to =  --------------------------------
+               max[Std[d field|on], min_scale]
+
+    Attributes:
+        to: name of the transformed variable.
+        condition_on: the variable to condition on
+        bins: the number of bins
+        field: the prognostic variable spec. The difference between input and
+            output data (``d field``) is normalized.
+        min_scale: the minimium scale to normalize by. Used when the scale might
+            be 0.
+
+    """
+
+    to: str
+    condition_on: str
+    bins: int
+    field: Field
+    min_scale: float = 0.0
```

Comment:
Would it be easier to just make the default some epsilon in the case of the scale being zero?

Reply:
It would be nice if the default worked for all variables---including ones with smaller non-zero magnitude than `min_scale`. I think the most sane default is to do nothing at all.